beetmoverscript/zip.py

```diff
@@ -0,0 +1,206 @@
+import logging
+import os
+import zipfile
+
+from scriptworker.exceptions import TaskVerificationError
+
+from beetmoverscript.constants import ZIP_MAX_COMPRESSION_RATIO
+
+log = logging.getLogger(__name__)
+
+
+def check_and_extract_zip_archives(artifacts_per_task_id, expected_files_per_archive_per_task_id, zip_max_size_in_mb):
+    deflated_artifacts = {}
+
+    for task_id, task_artifacts_params in artifacts_per_task_id.items():
+        for artifacts_param in task_artifacts_params:
+            if artifacts_param['zip_extract'] is False:
+                log.debug('Skipping artifacts marked as not `zipExtract`able: {}'.format(artifacts_param['paths']))
+                continue
+
+            expected_files_per_archive = expected_files_per_archive_per_task_id[task_id]
+            # No need to key deflated_artifacts by task_id. task_id is already in the full path of the archive
+            deflated_artifacts.update(_check_and_extract_zip_archives_for_given_task(
+                task_id, expected_files_per_archive, zip_max_size_in_mb
+            ))
+
+    log.info('Extracted these files: {}'.format(deflated_artifacts))
+
+    return deflated_artifacts
+
+
+def _check_and_extract_zip_archives_for_given_task(task_id, expected_files_per_archive, zip_max_size_in_mb):
+    extracted_files = {}
+
+    for archive_path, expected_files in expected_files_per_archive.items():
+        log.info('Processing archive "{}" which marked as `zipExtract`able'.format(archive_path))
+        extracted_files[archive_path] = _check_extract_and_delete_zip_archive(
+            archive_path, expected_files, zip_max_size_in_mb
+        )
+
+    # We make this check at this stage (and not when all files from all tasks got extracted)
+    # because files from different tasks are stored in different folders by scriptworker. Moreover
+    # we tested no relative paths like ".." are not used within the archive.
+    _ensure_no_file_got_overwritten(task_id, extracted_files)
+
+    return extracted_files
+
+
+def _check_extract_and_delete_zip_archive(zip_path, expected_files, zip_max_size_in_mb):
+    _check_archive_itself(zip_path, zip_max_size_in_mb)
+
+    with zipfile.ZipFile(zip_path) as zip_file:
+        zip_metadata = _fetch_zip_metadata(zip_file)
+
+        # we don't close the file descriptor here to avoid the tested file to be swapped by a rogue one
+        _ensure_files_in_archive_have_decent_sizes(zip_path, zip_metadata, zip_max_size_in_mb)
+        _ensure_all_expected_files_are_present_in_archive(zip_path, zip_metadata, expected_files)
+        log.info('Content of archive "{}" is sane'.format(zip_path))
+
+        extracted_files = _extract_and_check_output_files(zip_file, zip_metadata.keys())
```

Comment:
Nit: could be my own preferences so feel free to ignore this. Usually I avoid having magic logic happening in the params. I'd rather submit zip_metadata as dict here, and do that processing with keys extraction in the function itself.

Reply:
Yeah, I agree this is confusing. Thank you for spotting it. I fixed this in https://github.com/mozilla-releng/beetmoverscript/pull/163/commits/5d9fcf0caccb059410bfc7ddca9ecefae1a0cdb2 and https://github.com/mozilla-releng/beetmoverscript/pull/163/commits/86f9770b59d509a3ad11871354248b165b9122c3. Please tell me what you think!